[3D] correctly compute near/far planes when a scene is reloaded

ptitjano opened the pull request:
From the main commit:

```
The near/far planes are computed in `Qgs3DMapScene` when the camera
parameters have changed. Their values depend on the bounding boxes of
the visible entities. However, the near and far planes are not
recomputed when a new entity is added.
This can be problematic for entities from a vector layer because their
vertical extent is unknown when the associated bounding boxes are
created.

This issue is fixed by calling
`Qgs3DMapscene::updateCameraNearFarPlanes` when a new scene entity is
created. Indeed, on a new scene entity, the exact exact bounding box
has been computed by the loader of the chunked entity.
```

## Description

### Current behavior : the far plane is not correct when the scene is reloaded

[near_far.webm](https://github.com/qgis/QGIS/assets/497207/4d7edffa-2aff-4888-b2f1-343d97a313e7)


### Fixed behavior with this PR: far plane computation is correct


[near_far_after.webm](https://github.com/qgis/QGIS/assets/497207/fb029133-c7ba-4b0b-9839-df49780da33f)


cc @benoitdm-oslandia @wonder-sk @uclaros

ptitjano committed:
qgs3dmapcanvas: Update the navigation widget on near/far plane changes

`Qgs3DNavigationWidget` is updated when the camera parameters are
updated by `QgsCameraController`. However, the near far planes of the
camera can be updated from the `Qgs3DMapScene`. In that case, the
`Qgs3DNavigationWidget` also neeeds to be updated.

This issue is fixed by listening to the near/far plane changes from
the scene camera in `Qgs3DMapCanvas`.

nyalldawson commented:
Off topic, but do you notice the light intensity increase which happens when the settings are changed? I've tried to track this down and have had no luck... I suspect it's a qt3d internal bug. :cry:

wonder-sk commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -314,16 +314,14 @@ void Qgs3DMapScene::onCameraChanged()
-    // the active nodes may have changed as well
-    updateScene();
-    updateCameraNearFarPlanes();
-  }
```
I think this is not very clear to me. Before we had 1. updateScene(), 2. updateCameraNearFarPlanes(), then optionally followed by 1+2 again. This logic may be outdated now, after having everything handled as chunked entities, and with introduction of QgsChunkedEntity::getNearFarPlaneRange(), we can also account for entities with no active nodes).

Now we would have 1. updateCameraNearFarPlanes(), 2. updateScene(). I would agree with the new logic, that we first get approximate near/far planes and then update scene based on that. But why remove the final updateCameraNearFarPlanes() ? The updateScene() call may have activated some nodes, giving a chance for better near/far plane estimate...?

ptitjano replied:
Removed.

wonder-sk commented on tests/src/3d/testqgs3drendering.cpp:
```diff
@@ -1403,9 +1403,9 @@ void TestQgs3DRendering::testDepthBuffer()
-  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 126.4, 0.1 );
+  QGSCOMPARENEARVECTOR3D( testCam->zoomPoint(), QVector3D( -40.3, 276.8, 228.8 ), 1.0 );
+  QGSCOMPARENEARVECTOR3D( testCam->cameraPose()->centerPoint(), QVector3D( -40.3, 276.8, 228.8 ), 1.0 );
+  QGSCOMPARENEAR( testCam->cameraPose()->distanceFromCenterPoint(), 114.6, 0.1 );
```
why is there change in test results here, and in the test control images?

ptitjano replied:
Because the test was broken. The far plane taken into account was wrong.

ptitjano commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -893,7 +892,14 @@ void Qgs3DMapScene::setSceneState( Qgs3DMapScene::SceneState state )
+  {
+    // compute the correct near/far planes once all the entities are loaded
+    updateCameraNearFarPlanes();
+  }
```
I did some more testing and I've realized that the issue only occurs when the 3d scene contains vector layers. To reproduce the issue : 
- create a 3d scene with vector data and display the debug info to see the near/far plane values
- zoom in
- display the configuration dialog and click on ok => the near/far planes are computed
- the new near/far planes are completely off . For example, I get 150000 instead of 2000 for the near plane

You need the first commit from this PR, otherwise the near/far values are not updated in the debug panel

The problem comes from the vector layers. The initial vertical values of the bounding box are very important and therefore the initial guess of the near/far planes are completely off. By adding a `updateCameraNearFarPlanes` in once the scene is loaded, it ensures to get correct near/far plane values because the bounding box coordinates have been updated.

Do you think this issue could be solved with another approach?

wonder-sk replied:
How about if the near/far planes would be updated when we detect that some bounding boxes have been updated? If that's the reason why near/far plane should be changed, then we should react to that, rather than reacting when the whole scene got loaded?

ptitjano replied:
Thanks for the suggestion @wonder-sk . I have completely rewritten this PR based on your suggestion.

wonder-sk commented on src/app/3d/qgs3dmapcanvas.cpp:
```diff
@@ -119,6 +119,8 @@ void Qgs3DMapCanvas::setMap( Qgs3DMapSettings *map )
   connect( mScene, &Qgs3DMapScene::fpsCounterEnabledChanged, this, &Qgs3DMapCanvas::fpsCounterEnabledChanged );
   connect( mScene, &Qgs3DMapScene::viewed2DExtentFrom3DChanged, this, &Qgs3DMapCanvas::viewed2DExtentFrom3DChanged );
+  // update the navigation widget when the near far plane has been updated by the map scene
+  connect( mScene, &Qgs3DMapScene::nearFarPlaneChanged, mNavigationWidget, &Qgs3DNavigationWidget::updateFromCamera );
```
I don't think we need a new signal here. mScene->cameraController()->camera() returns Qt3DRender::QCamera, which has nearPlaneChanged() and farPlaneChanged() signals - let's use those!

ptitjano replied:
Done.

nyalldawson commented:
@ptitjano can you use test mask updates for the test images instead of replacing the reference image? There's only a handful of pixels different, so the mask should be used instead.

ptitjano committed:
qgs3dmapscene: Ensure to update near/far planes on new entity creation

The near/far planes are computed in `Qgs3DMapScene` when the camera
parameters have changed. Their values depend on the bounding boxes of
the visible entities. However, the near and far planes are not
recomputed when a new entity is added.
This can be problematic for entities from a vector layer because their
vertical extent is unknown when the associated bounding boxes are
created.

This issue is fixed by calling
`Qgs3DMapscene::updateCameraNearFarPlanes` when a new scene entity is
created. Indeed, on a new scene entity, the exact exact bounding box
has been computed by the loader of the chunked entity.

ptitjano commented:
> @ptitjano can you use test mask updates for the test images instead of replacing the reference image? There's only a handful of pixels different, so the mask should be used instead.

Done

wonder-sk commented:
@ptitjano please rebase...

ptitjano committed:
qgs3dmapcanvaswidget: Directly use canvas cameracontroller method

ptitjano commented:
> @ptitjano please rebase...

Done
